# Regular Expressions Lab

Les Carr, Last update 19/01/2023

Copyright University of Southampton, 2021.
Permission is granted for copies to be made for personal use by University of Southampton students. This content should not be shared on published outside the University of Southampton.

Regular expressions (aka regexes or regexps) are text matching patterns described with a formal syntax that were covered in Lecture 3. This lab covers two main issues
- Basic Python Capabilities for Using Regular Expressions
- How to use regular expressions for a variety of tasks


## Basic Python Capabilities for Regular Expressions

Regular expressions are implemented in the `re` package, which provides a number of functions to match regexps.


In [ ]:
import re

# List of patterns to search for

# Text to parse
text = 'This is a string with term1.'
if re.search("term[0-4]",  text):
        print('first term.')
if re.search("term[5-9]",  text):
        print ('Found second term.')

: 

Now we've seen that `re.search()` will take the pattern, scan the text, and then returns a **Match** object corresponding to the first match. If no pattern is found, a **None** is returned. Both of these are truthy, so any successful match (even a match on a null string from an ill-advised Kleene \*) will evaluate to `True` in a `Boolean` context. To give a clearer picture of this match object, check out the cell below:

In [3]:
match = re.search("\d+",  "This is the COMP3225 module's 1st lab")

print("The regexp matched '%s' between positions %s and %s" % (match.group(0), match.start(), match.end()))

The regexp matched '3225' between positions 16 and 20


If you used capture groups in the regular expression, they will appear as arguments `1` up to `99` of the `match.groups()` method. 

In [4]:
match = re.search("(\d+).*(\d+)",  "This is the COMP3225 module's 12st lab")

print("The regexp matched '%s' and '%s'" % (match.group(1), match.group(2)))

The regexp matched '3225' and '2'


There are three variants of the search function:
- `re.match()` is anchored at the beginning of the search string
- `re.fullmatch()` is anchored at the beginning and the end of the search string
- `re.findall()` return all matches 

You can also look for all the matches in a string with `re.findall()`, but it returns a list of the actual strings matched rather than a `Match` object.

In [6]:
match = re.findall("\d+",  "This is the COMP3225 module's 1st lab with 2nd lab")

print("The regexp matched '%s'" % (match))

The regexp matched '['3225', '1', '2']'


To embed this in a _file-read-and-match-print-results_ code fragment that works line by line, we can do the following. We are using the codecs package to explicitly manage the various character sets that we might encounter.  

In [8]:
import codecs
fname="../../corpus/comp3225/mytest.txt"

for line in codecs.open(fname,"r",encoding="utf-8"):
    match=re.findall("\d+", line)
    if match: print(match)

['1', '2021', '02380594479']
['19865']
['20', '49', '9']
['0', '10']


Finally, there are four usefull flags you can use to turn on different features in Python's implementation of regular expressions. 
- VERBOSE. Allow inline comments and extra whitespace.
- IGNORECASE. Do case-insensitive matches.
- DOTALL. Allow dot (.) to match any character, including a newline. (The default behavior of dot is to match anything, except for a newline.)
- MULTILINE. Allow anchors (^ and $) to match the beginnings and ends of lines instead of matching the beginning and end of the whole text.

They can be provided as flags to the re methods
- re.match("this", "This", flags=re.IGNORECASE)

or as abbreviated flags
- re.match("that", "That", flags=re.I)

or as inline flags in the regular expression itself
- re.match("(?i)those", "Those")


## Practicing Regular Expressions

You should now have a good enough understanding of how to use the regular expressions module in Python to undertake the rest of the tasks in the lab. You should refer to the full [Python regular expression documentation](https://docs.python.org/3/library/re.html)  for more detail.

Note that so far we haven't needed to use multiline or raw Python strings, or verbose regular expressions. As things get more complex, you might need to.

The following six tasks consist of three skills tasks and three application tasks. You should attempt all the skills tasks (1-3) to your satisfaction, and choose any of the application tasks (4-6) for general practice in using regular expressions to solve problems that require complex lexical analysis.

Please note: there are no right answers set for these tasks - no specific precision or recall targets that you have to meet. The intention is that you focus on practicing the use of regular expressions, and gain experience with applying different kinds of "patterns" to a range of situations.

## Task 1 - recognising specific lexical tokens
Recognise the following particular kinds of non-word lexical tokens, using your experience to define the allowable format of each token type. A whole load of data from the internet (including some HTML code) has been put into an example test file for you.

1. Hashtags (which can contain alphanumeric characters, underscores and hyphens). Test on the text of some tweets that you copy and paste from twitter.com.
1. UK Postcodes like SO17 1BJ or E1 8BN.
1. UK style phone numbers like 023 80591234 or 02380 590000 or +44 (0) 23 80594479 or +44 (23) 8059 5000 or mobile numbers like 07722 175921.
1. URLs http://google.com/ or https://a.b.net/c/d/e.php#fragment.
1. Email addresses like lac@ecs.soton.ac.uk, and decompose into username and internet domain.

In [14]:
import codecs
fname="../../corpus/comp3225/misctext.txt"
hashtag_re="#covid"
postcode_re="SO22 4NR"
phone_re="\+44 \(23\) 8059 5000"
countryRE = "([C][o][u][n][t][r][y][:])\s*([A-Z][a-z]+,(\s*[A-Z]*[a-z]+)+)"
#etc

for line in codecs.open(fname,"r",encoding="utf-8"):
    match=re.findall(hashtag_re, line)
    if match: print(match)
    match=re.findall(postcode_re, line)
    if match: print(match)
    match=re.findall(phone_re, line)
    if match: print(match)
    match=re.search(countryRE, line)
    if match: print(match)
    #etc

<re.Match object; span=(7, 30), match=': Bedfordshire, England'>
<re.Match object; span=(26, 41), match='y Park, Bedford'>
<re.Match object; span=(7, 27), match=': Yorkshire, England'>
<re.Match object; span=(13, 30), match='y Walk, Kirkstall'>
<re.Match object; span=(7, 32), match=': Leicestershire, England'>
<re.Match object; span=(19, 36), match='n Road, Leicester'>
<re.Match object; span=(13, 30), match='t Gallery, Kendal'>
<re.Match object; span=(7, 25), match=': Cumbria, England'>
<re.Match object; span=(18, 38), match='t Scotland, Scotland'>
<re.Match object; span=(7, 26), match=': Shiprow, Aberdeen'>
<re.Match object; span=(7, 29), match=': Oxfordshire, England'>
<re.Match object; span=(14, 31), match='t Place, Abingdon'>
<re.Match object; span=(7, 28), match=': Shropshire, England'>
<re.Match object; span=(7, 25), match=': Suffolk, England'>
<re.Match object; span=(16, 33), match='t Hall, Aldeburgh'>
<re.Match object; span=(35, 49), match='n Park, Newton'>
<re.Match object; sp

If that was too easy for you, look up the rules for the officially allowable formats of each token type in Wikipedia. Here for example, are the official UK postcode formats:

| Format	    | Coverage	| Example |
| -------:      | :--------:  | :------- |
| AA9A 9AA	| WC postcode area; EC1–EC4, NW1W, SE1P, SW1	| EC1A 1BB |
| A9A 9AA	| E1, N1, W1	| W1A 0AX |
| A9 9AA	| B, E, G, L, M, N, S, W	| M1 1AE |
| A99 9AA	| " | B33 8TH |
| AA9 9AA	| All other postcodes	| CR2 6XH  |
| AA99 9AA	| " | DN55 1PT |

The rules for email adddresses are eye-watering!

## Task 2 - numeric tokens

The zip file [guardian.zip](../corpus/guardian.zip) contains the text extracted from 118 Guardian news stories from the last year about Southampton, Portsmouth and Winchester. (One story per UTF-8 file, but you might like to combine them into a single file for ease of processing.) Starting from the regexp tokenizer example in Fig 2.12 (reproduced below), extend the set of tokens recognised to capture the following types of numeric data.
- Numbers: -12  47.2  74,832,101
- Time: 09:17pm
- Money: £27.8m £8bn
- Length: 6ft 48cm
- Phone: +44(0)2380594479
- Age specification: 13-year-old
- Percentage: 14.4%
- Temperature: 28C
- Ordinals: 48th 22nd 1st

You can compare your results with the [list of numeric tokens that I managed to extract](../corpus/guardian-numerics.txt). What’s the biggest financial quantity that appears in these stories? What did it relate to? What are the most common numeric tokens, and why do they appear?

In [ ]:
import nltk

text='That U.S.A. poster-print costs $12.40...'

# The following pattern is reproduced from the textbook figure 2.12.
# UNFORTUNATELY, the behaviour of NLTK has changed since version 3.1,
# so that capture groups don't work any more and every set
# of grouping parentheses () has now to explicitly declare non-capturing semantics with ?:
pattern = r'''(?x)			# set flag to allow verbose regexps 
	 (?:[A-Z]\.)+			# abbreviations, e.g. U.S.A. 
	 | \w+(?:-\w+)*			# words with optional internal hyphens 
	 | \$?\d+(?:\.\d+)?%?	# currency and percentages, e.g. $12.40, 82% 
	 | \.\.\.				# ellipsis 
	 | [][.,;"'?():-_`]		# these are separate punctuation tokens; includes ], [ 
	 '''

tokens=nltk.regexp_tokenize(text, pattern)
print(tokens)

## Task 3 - lookahead and lookbehind assertions
Use the lookaround capabilities to match the following (see the lookaround slide in lecture 3 as a start)
- a password with at least 6 characters, containing 2 upper case letters, two digits and two punctuation marks
- the above password, but it can't start with AB, ab, 01 or 12 to trap the really obvious abcd... and 1234... passwords
- separate out the different components of a camelCase word (e.g. getElementById -> [get, Element, by, Id])
-- Hint: identify each location where the previous character is lowercase and the next character is uppercase

In [ ]:
#code

## Optional Task 4 - Analysing Lightbulb Jokes
There were more than 10k lightbulb jokes told on Twitter in in 2015. Here are a couple of examples of the genre.
- Q: How many thriller writers does it take to change a lightbulb? A: Two. One to screw it almost all the way in and another to give a surprising twist at the end.
- Do you know how many folk musicians it takes to change a lightbulb? Five. One to change the lightbulb, and four to write songs about how much better the old bulb was.

The standard pattern for the opening of a lightbulb joke is _How many X does it take to change a lightbulb?_

The file [lightbulbs-2015.txt](../corpus/lightbulbs-2015.txt) contains one lightbulb joke tweet per line.
Write a python code that uses regular expressions to isolates the topic X of each of the jokes and use it to produce a summary of the top 100 topics of lightbulb humour.
- You should throw away the answers (punchlines) to the joke, just look for the topic
- Hint: you will need to use trial and error to deal with variations in language, case and punctuation. You may want to test sets of regular expressions in an editor such as VI to allow you to see the coverage and special cases before you commit them to python code.
- This is unmoderated Internet gathered data. Apologies for any inappropriate language that it might contain, or any examples of offensive humour. If you find anything that should be removed, please let me know.

The file [lightbulbs-2020.txt](../corpus/lightbulbs-2020.txt) contains the same data for last year. Use the regular expressions you developed to generate a top 100 summary for 2020 and compare the two years. What significant changes in topics have there been between 2015 and 2020?

In [ ]:
#download the file beforehand
fname="../../corpus/comp3225/lightbulbs-2015.txt"

#make the right regular expression to isolate just the topic
pattern="what (.*) what"

lineN=0; matchN=0
for line in codecs.open(fname,"r",encoding='utf-8'):
    lineN+=1
    match=re.search(pattern, line)
    if match:
        matchN+=1
        print(match.group(1))
print("The regexp matched %s jokes out of %s" % (matchN, lineN)) 

#now go on and do the analysis

## Optional Task 5 - citations

The file [thesis.txt](https://secure.ecs.soton.ac.uk/notes/comp3225/labs/citations-thesis.txt) contains the text of a PhD thesis. The examiner wants to know how many papers the student has cited, and of those, how many are “recent” (i.e. from the last decade).
The bibliographic style used means that citations appear in the text with the basic form (surnames year), and you can see examples of this as they appear in the PDF <img src="https://secure.ecs.soton.ac.uk/notes/comp3225/labs/citations-sample.png" width="400">

Write regular expressions to extract the list of citations from the thesis, answering the examiner's questions above.

In [ ]:
#code

### Optional Task 6 - ELIZA

ELIZA was an early natural language processing computer program created in 1964 (the year I was born) at the MIT Artificial Intelligence Laboratory and created to demonstrate superficial communication techniques (OMG - it's like we're twins).

Eliza simulated conversation by using a "pattern matching" and substitution methodology that gave users an illusion of understanding on the part of the program. Famously ELIZA simulated a style of psychotherapist, well-known for simply parroting back at patients what they had just said, and used rules to respond with non-directional questions to user inputs. As such, ELIZA was one of the first chatterbots and one of the first programs capable of attempting the Turing test.

You should implement an ELIZA-like conversational program, using substitutions such as those described on page 11 in the text book.

In [ ]:
#code